src/NHibernate/Dialect/MsSql2008Dialect.cs

```diff
@@ -36,5 +68,27 @@ protected override void RegisterDefaultProperties()
 			base.RegisterDefaultProperties();
 			DefaultProperties[Environment.ConnectionDriver] = typeof(Sql2008ClientDriver).AssemblyQualifiedName;
 		}
+
+		public override string CurrentTimestampSQLFunctionName =>
+			KeepDateTime ? base.CurrentTimestampSQLFunctionName : "SYSDATETIME()";
+
+		public override long TimestampResolutionInTicks =>
+			KeepDateTime
+				? base.TimestampResolutionInTicks
+				// MS SQL resolution with datetime2 is 100ns, one tick.
+				: 1;
+
+		/// <inheritdoc />
+		public override SqlType[] RefineSqlTypes(SqlType[] types)
+		{
+			types = base.RefineSqlTypes(types);
+
+			if (KeepDateTime)
+				return types;
+
+			return types
+				.Select(t => SqlTypeFactory.DateTime.Equals(t) ? SqlTypeFactory.DateTime2 : t)
+				.ToArray();
```

Comment:
This needs to be [`Array.ConvertAll`](https://msdn.microsoft.com/en-us/library/exc45z53)